app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt

```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun safeCameraOpen() {
+        val manager : CameraManager = getSystemService(Context.CAMERA_SERVICE) as CameraManager
+        try {
+            val cameraId = manager.cameraIdList[0]
+            val characteristics = manager.getCameraCharacteristics(cameraId)
+            val map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
+            mPreviewSize = map.getOutputSizes(SurfaceTexture::class.java)[0]
+            manager.openCamera(cameraId, mStateCallback, null)
+        }
+        catch (e: CameraAccessException) {
+            e.printStackTrace()
+        }
+    }
+
+    private var mStateCallback: CameraDevice.StateCallback = object: CameraDevice.StateCallback() {
+        override fun onOpened(camera: CameraDevice?) {
+            Log.i(TAG, "onOpened")
+            mCameraDevice = camera
+            val texture = mTextureView.surfaceTexture
+            if (texture == null) {
+                Log.e(TAG, "texture is null")
+                return
+            }
+
+            texture.setDefaultBufferSize(mPreviewSize.width, mPreviewSize.height)
+            val surface = Surface(texture)
+
+            try {
+                mPreviewBuilder = mCameraDevice!!.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
+            } catch (e: CameraAccessException) {
+                e.printStackTrace()
+            }
+
+
+            mPreviewBuilder.addTarget(surface)
+
+            try {
+                mCameraDevice!!.createCaptureSession(arrayListOf(surface), mPreviewStateCallback, null)
+            } catch (e: CameraAccessException) {
+                e.printStackTrace()
+            }
+
+        }
+
+        override fun onDisconnected(camera: CameraDevice?) {
+            //TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+            Log.e(TAG, "onDisconnected");
+        }
+
+        override fun onError(camera: CameraDevice?, int: Int) {
+            //TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+            Log.e(TAG, "onError")
+        }
+    }
+
+    private val mPreviewStateCallback = object : CameraCaptureSession.StateCallback() {
+
+        override fun onConfigured(session: CameraCaptureSession) {
+            // TODO Auto-generated method stub
+            Log.i(TAG, "onConfigured")
+            mPreviewSession = session
+
+            mPreviewBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO)
+
+            val backgroundThread = HandlerThread("CameraPreview")
+            backgroundThread.start()
+            val backgroundHandler = Handler(backgroundThread.looper)
+
+            try {
+                mPreviewSession.setRepeatingRequest(mPreviewBuilder.build(), null, backgroundHandler)
+            } catch (e: CameraAccessException) {
+                e.printStackTrace()
```

Comment:
I think we should display a Toast message here to notify the user.

Reply:
notify about what? the exception?